src/main/java/emissary/place/CoordinationPlace.java

```diff
@@ -239,30 +232,108 @@ protected List<IBaseDataObject> coordinate(IBaseDataObject d, boolean hd) {
-                } else {
-                    d.setCurrentForm(outputForm);
-                }
-            }
+        applyForm(d, errorOccurred);
 
-            // Clean up my proxies
-            nukeMyProxies(d);
+        sproutHook(d, hd, sproutCollection);
 
-        }
+        // Allow derived classes a shot to clean up the parent
+        cleanUpHook(d);
+
+        return sproutCollection;
+    }
 
+    private void sproutHook(IBaseDataObject d, boolean hd, List<IBaseDataObject> sproutCollection) {
         // Allow derived classes a shot at the sprouts
         if (hd) {
             sproutHook(sproutCollection, d);
         }
+    }
 
-        // Allow derived classes a shot to clean up the parent
-        cleanUpHook(d);
+    /**
+     * Allow derived classes a shot to handle a place exception
+     *
+     * @param p place that was processing when the exception was thrown
+     * @param hd true if doing heavy-duty processing
+     * @param ex exception thrown by the place
+     * 
+     * @return if an error occurred
+     */
+    protected boolean handledPlaceException(IServiceProviderPlace p, boolean hd, Exception ex) {
+        logger.warn("agentProcess{} called from Coordinate problem", (hd ? "HeavyDuty" : "Call"), ex);
+        return true;
+    }
 
-        return sproutCollection;
+    /**
+     * Return false when an error has occurred and the places is configured to not continue
+     *
+     * @param p place that is currently processing the ibdo
+     * @param errorOccurred true if an error occurred
+     *
+     * @return false if processing should not continue
+     */
+    protected boolean shouldContinue(IServiceProviderPlace p, boolean errorOccurred) {
+        if (!continueOnError() && errorOccurred) {
+            logger.info("Error terminating coordination step at {}", p);
+            return false;
+        }
+        return true;
+    }
+
+    private void updateTransformHistory(IBaseDataObject d, IServiceProviderPlace p) {
+        if (updateTransformHistory) {
+            DirectoryEntry de = p.getDirectoryEntry();
+            de.setDataType(d.currentForm());
+            // append to the transform history, with flag indicating that the visit was coordinated
+            d.appendTransformHistory(de.getKey(), true);
+        }
+    }
+
+    /**
+     * Apply the form
+     *
+     * @param d the ibdo to process
+     * @param errorOccurred true if an error occurred
+     */
+    protected void applyForm(IBaseDataObject d, boolean errorOccurred) {
+        if (!errorOccurred || shouldApplyOutputFormOnError()) {
+            applyOutputForm(d);
+
+            // Clean up my proxies
+            nukeMyProxies(d);
+        }
+    }
+
+    /**
+     * If true, process the output form the same for the default and error condition
+     *
+     * @return boolean to allow processing of output for when an error occurs
+     */
+    protected boolean shouldApplyOutputFormOnError() {
```

Comment:
(optional) Could we move this just before the other `shouldXXX` methods?

Reply:
Scratch that - this is already close to where it's used so the cohesion is still pretty good